distributed/scheduler.py

```diff
@@ -4620,6 +4592,42 @@ def update_graph(
 
         # TODO: balance workers
 
+    def stimulus_queue_slots_maybe_opened(self, *, stimulus_id: str) -> None:
+        """Respond to an event which may have opened spots on worker threadpools
+
+        Selects the appropriate number of tasks from the front of the queue according to
+        the total number of task slots available on workers (potentially 0), and
+        transitions them to ``processing``.
+
+        Notes
+        -----
+        Other transitions related to this stimulus should be fully processed beforehand,
+        so any tasks that became runnable are already in ``processing``. Otherwise,
+        overproduction can occur if queued tasks get scheduled before downstream tasks.
+
+        Must be called after `check_idle_saturated`; i.e. `idle_task_count` must be up
+        to date.
+        """
+        if not self.queued:
+            return
+        slots_available = sum(
+            _task_slots_available(ws, self.WORKER_SATURATION)
+            for ws in self.idle_task_count
+        )
```

Comment:
This is hardly worth mentioning, but potential for a negligible scheduler performance change here (which may not actually be possible):

Now, every time a task completes, we run `_task_slots_available` on all idle workers. Before, we only ran it on the one worker that just completed a task. In most of those cases, `len(self.idle_task_count)` should be 1, so no difference.

I mention this because `_task_slots_available` does already show up in py-spy profiles of the scheduler (usually around 0.5-1%), because it's already called frequently. This would maybe allow it to be called even more frequently.

But I don't think it can actually run more than it needs to:
* When there are more threads than root tasks, so many workers are idle, `queued` would be empty, so this wouldn't run.
* When there are more root tasks than threads, and we're queuing, at most one worker should ever be idle: as soon as it becomes idle, it gets another queued task and is no longer idle.

    The one exception is `client.close()` releasing many processing tasks at once, while a different client has tasks on the queue. In that case, there's a lot of rescheduling to do, and we do need to look at all the workers that just became idle, so no unnecessary work there either.

tl;dr I don't see a theoretical way for this to be a problem, but I haven't benchmarked or profiled to be sure.

Reply:
Thanks for highlighting.

Apart from the theoretical analysis, this is also something we can easily optimize if it becomes a problem, e.g. `saturation * nthreads` is not something we need to compute every time and even the `processing - long running` could be replaced with a counter we inc/dec during transitions.

TLDR Not concerned and "fixing" it right now feels premature